### PR TITLE
Add order refunds (SHUUP-2400, SHUUP-2821, SHUUP-2874, SHUUP-2875, SHUUP-2877)

### PR DESCRIPTION
**Pull request by shawnadelic**
- Add core support for refunds
- Add refund creator in Orders admin
- Show log entries section or all orders

Refs SHUUP-2400, SHUUP-2821, SHUUP-2874, SHUUP-2875, SHUUP-2877

### CODE REVIEW

File: shoop/admin/modules/orders/views/refund.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+        required=False,
+        min_value=0,
+        initial=0,
+        label=_("Amount"),
+    )
+    restock_products = forms.BooleanField(
+        required=False,
+        initial=True,
+        label=_("Restock products")
+    )
```
Comment (Pikkupomo) [inline shoop/admin/modules/orders/views/refund.py]: these have way too many line changes. I think these can be fitted in one row, all of these fields.

File: shoop/core/models/_orders.py [hunk 2]
```diff
@@ -514,6 +515,100 @@ def create_shipment(self, product_quantities, supplier=None, shipment=None):
...
+        :param restock_products: Boolean indicating whether to restock products
+        :type restock_products: bool|False
+        """
+        if self.has_refunds():
+            raise NoRefundToCreateException
+        self.cache_prices()
+        amount = self.taxful_total_price
+        self.create_refund([{"amount": amount.value}])
+        if restock_products:
+            for product_line in self.lines.filter(type=OrderLineType.PRODUCT):
```
Comment (Pikkupomo) [inline shoop/core/models/_orders.py]: Could this work: `for product_line in self.lines.filter(type=OrderLineType.PRODUCT, product__stock_behavior=StockBehavior.STOCKED)`

File: shoop/core/models/_orders.py [hunk 3]
```diff
@@ -514,6 +515,100 @@ def create_shipment(self, product_quantities, supplier=None, shipment=None):
...
+                if product_line.product.stock_behavior == StockBehavior.STOCKED:
+                    product_line.supplier.adjust_stock(product_line.product.id, product_line.quantity)
+
+    def get_total_refunded_amount(self):
+        total = sum(line.taxful_price.amount.value for line in self.lines.filter(type=OrderLineType.REFUND))
```
Comment (Pikkupomo) [inline shoop/core/models/_orders.py]: Perhaps you could use the manager you added here? `self.lines.refunds()`

File: shoop/core/models/_orders.py [hunk 4]
```diff
@@ -514,6 +515,100 @@ def create_shipment(self, product_quantities, supplier=None, shipment=None):
...
+        return Money(total * (-1), self.currency)
+
+    def get_total_unrefunded_amount(self):
+        return max(self.taxful_total_price.amount, Money(0, self.currency))
+
+    def has_refunds(self):
+        return self.lines.filter(type=OrderLineType.REFUND).exists()
```
Comment (Pikkupomo) [inline shoop/core/models/_orders.py]: Perhaps you could use the manager you added here? `self.lines.refunds().exists()`

File: shoop/core/models/_orders.py [hunk 5]
```diff
@@ -514,6 +515,100 @@ def create_shipment(self, product_quantities, supplier=None, shipment=None):
...
+
+    def can_create_refund(self):
+        return self.taxful_total_price.amount.value > 0
```
Comment (Pikkupomo) [inline shoop/core/models/_orders.py]: `return (self.taxful_total_price.amount.value > 0)` here please

File: shoop/core/models/_order_lines.py [hunk 6]
```diff
@@ -106,6 +111,12 @@ def tax_amount(self):
         zero = Money(0, self.order.currency)
         return sum((x.amount for x in self.taxes.all()), zero)
 
+    @property
+    def max_refundable_amount(self):
+        refunds = OrderLine.objects.refunds().filter(order=self.order, parent_line=self)
+        refund_total = sum(refund.taxful_price.amount.value for refund in refunds)
+        return self.taxful_price.amount.value - (refund_total * (-1))
```
Comment (Pikkupomo) [inline shoop/core/models/_order_lines.py]: This should return `Money`: `return (self.taxful_price.amount + refund_total)`

File: shoop/core/models/_orders.py [hunk 7]
```diff
@@ -514,6 +515,100 @@ def create_shipment(self, product_quantities, supplier=None, shipment=None):
...
+            unit_price = parent_line.base_unit_price.amount.value if parent_line else 0
+            total_price = unit_price * quantity + amount_value
+
+            refund_line = OrderLine.objects.create(
+                text=_("Refund for %s" % parent_line.text) if parent_line else _("Manual refund"),
+                order=self,
+                type=OrderLineType.REFUND,
+                parent_line=parent_line,
+                ordering=index,
+                base_unit_price_value=Money(total_price * (-1), self.currency),
```
Comment (Pikkupomo) [inline shoop/core/models/_orders.py]: Aren't values usually "raw" Decimals?

File: shoop/core/models/_order_lines.py [hunk 8]
```diff
@@ -106,6 +111,16 @@ def tax_amount(self):
         zero = Money(0, self.order.currency)
         return sum((x.amount for x in self.taxes.all()), zero)
 
+    @property
+    def max_refundable_amount(self):
+        """
+        :rtype: shoop.utils.money.Money
+        """
+        refunds = OrderLine.objects.refunds().filter(order=self.order, parent_line=self)
+        refunds = self.order.lines.refunds().filter(parent_line=self)
```
Comment (Pikkupomo) [inline shoop/core/models/_order_lines.py]: refunds is set twice

File: shoop/core/excs.py [hunk 9]
```diff
@@ -20,6 +20,14 @@ class NoPaymentToCreateException(Exception):
     pass
 
 
+class NoRefundToCreateException(Exception):
+    pass
+
+
+class RefundExceedsAmountException(Exception):
+    pass
```
Comment (suutari-ai) [inline shoop/core/excs.py]: These should probably inherit from ValueError

**Pull request merged**